### DISCUSSION TITLE
Ractor safety

### DISCUSSION THREAD

**Discussion by HoneyryderChuck** [jeremyevans/sequel#1895] [General]
👋 I am using sequel for this tool I'm working on, where there's a pool of workers, each using the default connection pool to pick a connection tu use. Recently, I've been toying with the idea of adding support for a pool of ractors. In order to make it work though, sequel would need to at least be ractor-safe, which is not at the moment.

The first issue is that database instances, even if frozen, aren't shareable:

```ruby
DB = Sequel.connect("...")

Ractor.new(DB) do |db|
   # does not work, because Sequel::Database does not implement #initialize_copy
```

This can be overcome by initializing the database instance within the ractor:

```ruby


Ractor.new do
  db = Sequel.connect("...")
```

But then all other sorts of issues happen, the first of which is, [using ADAPTER_MAP](https://github.com/jeremyevans/sequel/blob/master/lib/sequel/database/connecting.rb#L77) when connecting fails, as it's a mutable constant.

I've made a local patch to ditch its usage and rely on `Sequel::Database.subclasses.find` and `.adapter_scheme`. I believe the scheme cache of `ADAPTER_MAP` optimizes a cold path (connecting happens usually once per database URL, and there aren't usually many of them). There's of course the open question of what to do with the [double declaration of the postgres adapter](https://github.com/jeremyevans/sequel/blob/master/lib/sequel/adapters/postgres.rb#L167) overriding the scheme, I guess this is for backwards-compatibility?

And this is just the beginning: `Sequel.synchronize` is used liberally in a few places, and is definitely not ractor-friendly; moreover, `sequel` caches a lot of [method-to-proc coercers](https://github.com/jeremyevans/sequel/blob/7c3a0aa75471ea55809ac9cf95c8cf950d76a44c/lib/sequel/adapters/shared/postgres.rb#L67), which aren't themselves ractor-shareable [I've created an issue in ruby about whether these procs should be considered exceptions](https://bugs.ruby-lang.org/issues/18919).

--- 

This is the summary of early discovery. I don't know how much Jeremy is interested in making  `sequel` work under ractors, and while I believe that it's going to take a lot of effort making the whole of `sequel` ractor-friendly, I wonder how much one has to do in order to make the vanilla "no-plugin" XP usable.

Is this something worth undertaking, or a rabbit-hole someone may have gone through in the past already?

**Comment by jeremyevans** [comment]
With current limitations of Ractor, I don't think there is any chance of Sequel being able to use it.  If you want thread-level parallelism when using Sequel, I recommend using JRuby.
